apolline-flutter/lib/twins/SensorTwin.dart

```diff
@@ -218,16 +219,37 @@ class SensorTwin {
   DataPointModel _handleSensorUpdate (String message) {
     if (!message.contains('\n')) return null;
     print("Got full line: " + message);
-    List<String> values = message.split(';');
-
-    var model = DataPointModel(values: values, sensorName: this.name, position: _currentPosition);
+    DataPointModel model = this._getPointWithPosition(message.split(';'));
     _dataService.update(model);
     /* insert to sqflite */
     _sqfLiteService.addDataPoint(model.toJSON());
 
     return model;
   }
 
+  /// Returns a data point with the current location.
+  /// Current location is either:
+  ///   * sensor location, if it currently has access to GPS signal;
+  ///   * phone location otherwise.
+  DataPointModel _getPointWithPosition (List<String> values) {
+    double sensorLongitude = double.parse(values[DataPointModel.SENSOR_LONGITUDE]);
+    double sensorLatitude = double.parse(values[DataPointModel.SENSOR_LATITUDE]);
+
+    Position currentPosition;
+
+    if (sensorLongitude == 0 && sensorLatitude == 0) {
+      currentPosition = _currentPosition;
+      _initLocationService();
+    } else {
+      currentPosition = Position(
+          provider: "sensor",
+          geohash: SimpleGeoHash.encode(sensorLatitude, sensorLongitude));
```

Comment:
Check if sensor GPS accuracy can be increased.

Reply:
It can't, so we have to look at satellites count to decide whether to use phone or sensor GPS.